Build Failure: `biopython`

### Steps To Reproduce

Minimal flake to reproduce the issue:

`flake.nix`:

```nix
{
  inputs.nixpkgs.url = "github:NixOS/nixpkgs/nixpkgs-unstable";

  outputs = {
    self,
    nixpkgs,
  }: let
    forAllSystems = nixpkgs.lib.genAttrs [
      "aarch64-linux"
      "aarch64-darwin"
      "x86_64-darwin"
      "x86_64-linux"
    ];
    mkPkgs = system:
      import nixpkgs {
        inherit system;
        overlays = [
          # Uncomment this overlay and the issue will be fixed
          # (final: prev: {
          #   python3 = prev.python3.override {
          #     packageOverrides = pyfinal: pyprev: {
          #       biopython = pyprev.biopython.overrideAttrs (_: {
          #         doInstallCheck = false;
          #       });
          #     };
          #   };
          # })
        ];
      };
  in {
    # Development environment
    # Acessible through `nix develop`
    devShells = forAllSystems (system: let
      pkgs = mkPkgs system;
    in {
      default = pkgs.mkShell {
        nativeBuildInputs = [(pkgs.python3.withPackages (ps: [ps.biopython]))];
      };
    });
  };
}
```

`flake.lock`:

```nix
{
  "nodes": {
    "nixpkgs": {
      "locked": {
        "lastModified": 1708976803,
        "narHash": "sha256-yvRygcySjjSvj5JTaCdo7lPqJ/2mBV2XQ94Oaq/14qw=",
        "owner": "NixOS",
        "repo": "nixpkgs",
        "rev": "548a86b335d7ecd8b57ec617781f5e652ab0c38e",
        "type": "github"
      },
      "original": {
        "owner": "NixOS",
        "ref": "nixpkgs-unstable",
        "repo": "nixpkgs",
        "type": "github"
      }
    },
    "root": {
      "inputs": {
        "nixpkgs": "nixpkgs"
      }
    }
  },
  "root": "root",
  "version": 7
}
```

### Build log

[fail.log](https://github.com/NixOS/nixpkgs/files/14438672/fail.log)

### Additional context

Disabling install checks using `doInstallCheck = false` fixes the issue. If you uncomment the commented overlay in the minimal flake I provided, the build will succeed and the package will work properly.

### Notify maintainers

@luispedro 

### Metadata

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.74, NixOS, 24.05 (Uakari), 24.05.20240125.ae5c332`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - nixpkgs: `/nix/store/9s5qs4hni9fj88x79iw6im7amv7ghb76-source`
```

Hi, I tracked this issue down to a libexpat.so.1 upgrade from 2.5.0 to 2.6.1 from Debian testing to sid.  [Debian CI] suggests this could also cause regression in Python tests.  Is this consistent with what you observe on Nix?

If so, I strongly believe the root cause of the Biopython issue is ingrained in a bad interaction between the newer expat version and the Python interpreter, notably:
```
2510s ======================================================================
2510s FAIL: test_simple_xml (test.test_xml_etree.XMLPullParserTest.test_simple_xml) (chunk_size=1)
2510s ----------------------------------------------------------------------
2510s Traceback (most recent call last):
2510s   File "/usr/lib/python3.11/test/test_xml_etree.py", line 1415, in test_simple_xml
2511s     self.assert_event_tags(parser, [('end', 'element')])
2511s   File "/usr/lib/python3.11/test/test_xml_etree.py", line 1400, in assert_event_tags
2511s     self.assertEqual([(action, elem.tag) for action, elem in events],
2511s AssertionError: Lists differ: [] != [('end', 'element')]
2511s 
2511s Second list contains 1 additional elements.
2511s First extra element 0:
2511s ('end', 'element')
2511s 
2511s - []
2511s + [('end', 'element')]
2511s 
2511s ======================================================================
2511s FAIL: test_simple_xml (test.test_xml_etree.XMLPullParserTest.test_simple_xml) (chunk_size=5)
2511s ----------------------------------------------------------------------
2511s Traceback (most recent call last):
2511s   File "/usr/lib/python3.11/test/test_xml_etree.py", line 1418, in test_simple_xml
2511s     self.assert_event_tags(parser, [
2511s   File "/usr/lib/python3.11/test/test_xml_etree.py", line 1400, in assert_event_tags
2511s     self.assertEqual([(action, elem.tag) for action, elem in events],
2511s AssertionError: Lists differ: [('end', 'element')] != [('end', 'element'), ('end', 'empty-element')]
2511s 
2511s Second list contains 1 additional elements.
2511s First extra element 1:
2511s ('end', 'empty-element')
2511s 
2511s - [('end', 'element')]
2511s + [('end', 'element'), ('end', 'empty-element')]
```

[Debian CI]: https://ci.debian.net/packages/p/python3.11/testing/amd64/43739090/

In hope this helps,
Étienne.

--- a/pkgs/development/python-modules/biopython/close_parser_on_time.patch
+++ b/pkgs/development/python-modules/biopython/close_parser_on_time.patch
@@ -1,0 +1,18 @@
+diff --git a/Bio/SeqIO/SeqXmlIO.py b/Bio/SeqIO/SeqXmlIO.py
+index 8fe75ebb728..6758317d05f 100644
+--- a/Bio/SeqIO/SeqXmlIO.py
++++ b/Bio/SeqIO/SeqXmlIO.py
+@@ -498,11 +498,12 @@ def iterate(self, handle):
+             if not text:
+                 break
+             parser.feed(text)
++        # Closing the parser ensures that all XML data fed into it are processed
++        parser.close()
+         # We have reached the end of the XML file;
+         # send out the remaining records
+         yield from records
+         records.clear()
+-        parser.close()
+ 
+ 
+ class SeqXmlWriter(SequenceWriter):

--- a/pkgs/development/python-modules/biopython/default.nix
+++ b/pkgs/development/python-modules/biopython/default.nix
@@ -18,6 +18,13 @@ buildPythonPackage rec {
     hash = "sha256-eOa/t43mMDQDev01/nfLbgqeW2Jwa+z3in2SKxbtg/c=";
   };
 
+  patches = [
+    # cherry-picked from https://github.com/biopython/biopython/commit/3f9bda7ef44f533dadbaa0de29ac21929bc0b2f1
+    # fixes SeqXMLIO parser to process all data. remove on next update
+    ./close_parser_on_time.patch
+  ];
+
+
   nativeBuildInputs = [
     setuptools
   ];

--- a/pkgs/development/python-modules/biopython/default.nix
+++ b/pkgs/development/python-modules/biopython/default.nix
@@ -1,9 +1,10 @@
-{ lib
-, buildPythonPackage
-, fetchPypi
-, pythonOlder
-, setuptools
-, numpy
+{
+  lib,
+  buildPythonPackage,
+  fetchPypi,
+  pythonOlder,
+  setuptools,
+  numpy,
 }:
 
 buildPythonPackage rec {
@@ -24,16 +25,11 @@ buildPythonPackage rec {
     ./close_parser_on_time.patch
   ];
 
+  build-system = [ setuptools ];
 
-  nativeBuildInputs = [
-    setuptools
-  ];
+  dependencies = [ numpy ];
 
-  propagatedBuildInputs = [ numpy ];
-
-  pythonImportsCheck = [
-    "Bio"
-  ];
+  pythonImportsCheck = [ "Bio" ];
 
   checkPhase = ''
     runHook preCheck

